### PR TITLE
Preserve forward AD tangents for Inductor dual inputs

### LINKED ISSUE #187284: `torch.compile` with `inductor` backend silently drops forward-mode AD tangents (returns `None`)
### 🐛 Describe the bug

When a function is compiled with `torch.compile(backend="inductor")` and called inside a `torch.autograd.forward_ad.dual_level()` context, the forward-mode AD tangent (JVP) is silently dropped — `unpack_dual(...).tangent` returns `None` instead of the expected tangent tensor.

The same function produces the correct tangent in eager mode and with the `cudagraphs` backend.

## Minimal reproduction

```python
import torch
import torch.autograd.forward_ad as fwAD

def fn(x):
    return (x ** 2).sum()

compiled_fn_inductor = torch.compile(fn, backend="inductor")
compiled_fn_cudagraphs = torch.compile(fn, backend="cudagraphs")

x = torch.tensor([0.1, 0.2, 0.3])
v = torch.ones(3)

with fwAD.dual_level():
    dual_input = fwAD.make_dual(x, v)
    out = compiled_fn_inductor(dual_input)
    jvp_inductor = fwAD.unpack_dual(out).tangent

with fwAD.dual_level():
    dual_input = fwAD.make_dual(x, v)
    out = compiled_fn_cudagraphs(dual_input)
    jvp_cudagraphs = fwAD.unpack_dual(out).tangent

with fwAD.dual_level():
    dual_input = fwAD.make_dual(x, v)
    out = fn(dual_input)
    jvp_eager = fwAD.unpack_dual(out).tangent

print("inductor compiled jvp:", jvp_inductor)
print("cudagraphs compiled jvp:", jvp_cudagraphs)
print("eager jvp:", jvp_eager)
```

## Output

```
inductor compiled jvp: None
cudagraphs compiled jvp: tensor(1.2000)
eager jvp: tensor(1.2000)
```

### Versions

Collecting environment information...
PyTorch version: 2.11.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 26.5.1 (arm64)
GCC version: Could not collect
Clang version: 16.0.6
CMake version: version 4.0.3
Libc version: N/A

Python version: 3.10.19 (main, Oct 21 2025, 16:37:10) [Clang 20.1.8 ] (64-bit runtime)
Python platform: macOS-26.5.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Apple M3

Versions of relevant libraries:
[pip3] numpy==2.2.6
[pip3] optree==0.19.0
[pip3] torch==2.11.0
[pip3] torchvision==0.26.0
[conda] numpy                     2.2.6                    pypi_0    pypi
[conda] optree                    0.19.0                   pypi_0    pypi
[conda] torch                     2.11.0                   pypi_0    pypi
[conda] torchvision               0.26.0                   pypi_0    pypi

cc @ezyang @gchanan @kadeng @msaroufim @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @muchulee8 @amjames @aakhundov @coconutruben @jataylo @azahed98 @bdhirsh @bobrenjc93 @aorenste @Chillee @samdow @kshitij12345

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.16.0) (oldest at bottom):
* __->__ #187311

Inductor generated callables compute primal tensors only. When a compiled
callable is invoked with dual tensor inputs, the generated runtime bypasses
normal dispatcher JVP propagation and silently returns an output without a
forward AD tangent. Dynamo guards the active dual level, but that does not
prevent a compiled graph from being executed with tangent-bearing inputs.

Fix this by wrapping CompiledFxGraph callables with a runtime forward-AD check.
When boxed inputs carry active forward gradients, the wrapper executes the
original FX graph instead of the generated callable, preserving tangents through
normal eager dispatcher rules. Cache artifacts now preserve the original FX
graph and lazily deserialize it only when a forward-AD input reaches a cache-hit
path; normal cache hits do not load the graph. A cache-key bit prevents reuse of
older entries that were saved without this fallback graph.

The hessian example in the issue is a separate double-backward limitation of
AOTAutograd: direct double backward through a compiled function raises that this
is unsupported, and the issue's hessian repro also fails with cudagraphs. This
change targets the Inductor-specific forward AD tangent loss.

Fixes #187284
Generated by my agent

Benchmark Results:
- Command: python /tmp/bench_187284.py
- Workload: 20 repeated FX graph cache hits for a small CPU Inductor graph.
- Baseline clean worktree: median_ms 26.183, mean_ms 38.392, cache_bytes 408348
- Patched: median_ms 23.849, mean_ms 31.93, cache_bytes 408348
- Normal cache-hit check: GraphPickler.loads count remained 0 without forward AD inputs; forward-AD cache hit lazily loaded once and returned tangent tensor(1.2000).

Test Plan:
- python test/dynamo/test_higher_order_ops.py -k test_forward_ad_dual_input_inductor
- python test/dynamo/test_higher_order_ops.py FuncTorchHigherOrderOpTests.test_jvp_call_torch_compile_fn FuncTorchHigherOrderOpTests.test_forward_ad_dual_input_inductor
- python test/inductor/test_compiled_fx_graph_serialization.py -v
- TORCHINDUCTOR_FX_COMPILE_MODE=SERIALIZE python manual regular-cache forward-AD repro
- TORCHINDUCTOR_FX_COMPILE_MODE=SERIALIZE python manual bundled-autograd-cache forward-AD repro
- lintrunner -a

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @azahed98

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I addressed the review feedback by avoiding full GraphModule retention: the runtime wrapper no longer captures constants.gm, it keeps only a metadata-stripped fallback graph, installs the forward-AD fallback only for forward graphs, and disables bundled autograd cache serialization from Dynamo tracing. I also removed the unused prepare_for_serialization flag and added the missing release notes label.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I addressed the latest review feedback and CI-risk areas by breaking the forward-AD fallback self-reference cycle with a weakref and adding a frozen-parameter FX graph cache-hit forward-AD test. The targeted tests now cover the original tangent repro, serialized CompiledFxGraph fallback, bundled autograd cache, invoke-subgraph cache, and FX graph cache freezing behavior; lintrunner -a passes.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I updated the PR to address the new Inductor CI failures. The fallback graph now avoids deep-copying frozen/opaque tensor storage, reattaches frozen constants for eager fallback execution, prevents wrap_inductor_compiled_regions from recursively calling the forward-AD fallback, and skips FX graph cache writes if preparing the preserved fallback graph is unpicklable. I added a CPU regression for wrap_inductor_compiled_regions with forward AD dual inputs and reran the targeted failing tests plus lintrunner.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I updated the PR for the new CUDA opinfo CI failure. The forward-AD fallback probe now skips unpack_dual while inference mode is enabled, which avoids the _fw_primal internal assert when a test has an active dual level but non-dual inference-mode inputs. I added a regression covering active forward-AD level plus inference_mode and reran the reported CUDA opinfo, AOTInductor logging test, forward-AD fallback tests, serialization tests, wrap_inductor_compiled_regions coverage, and lintrunner.

**Comment by jansel** [comment]
My agent says I updated the PR for the remaining CUDA opinfo CI failure. The forward-AD probe now catches only the internal _fw_primal RuntimeError raised when AOTAutograd custom-op aliasing analysis routes unpack_dual through an internal inference-mode path, then declines the fallback for that wrapper-only case while re-raising other unpack_dual errors. I reproduced and reran TestCompositeComplianceCUDA.test_forward_ad_linalg_solve_triangular_cuda_float32 under PYTORCH_TEST_WITH_INDUCTOR=1, the forward-AD Inductor regressions, wrap_inductor_compiled_regions coverage, serialization tests, and lintrunner.

**Comment by jansel** [comment]
My agent says #189644 closed the linked issue for the direct dual-input compile path, but current main still drops the tangent when a function is first compiled with a plain tensor and later called with a forward-AD dual tensor from the cache-hit path. I verified on current main: direct dual input returns tangent tensor(1.2000), while plain-first then dual input still returns tangent None. This PR remains relevant for that runtime cache-hit case.

File: torch/_inductor/output_code.py [hunk 1]
```diff
@@ -913,32 +1019,30 @@ def post_compile(
...
-            fake_mode = FakeTensorMode(
-                allow_non_fake_inputs=True,
-                shape_env=ShapeEnv(),
-            )
-            original_gm = cast(
-                torch.fx.GraphModule,
-                GraphPickler.loads(self._serialized_original_gm, fake_mode),
+        if (
+            self._original_gm is None
+            and self._wrap_compiled_regions
```
Comment (bobrenjc93) [inline torch/_inductor/output_code.py]: The rebuild of `_original_gm` from `constants.gm` is gated on `self._wrap_compiled_regions`, but the forward-AD fallback below is installed unconditionally. So on a plain (non-wrap) cache hit the fallback uses `_load_original_gm()`, i.e. the graph pickled at save time, whose `get_attr` frozen params/constants are the *saved* tensors, not the current module's. With `freezing=True` a cache hit from a second module instance whose weights differ will compute the compiled result from the new weights but the tangent from the old ones (silently wrong). Either apply the `constants.gm` rebuild whenever `isinstance(constants, CompiledFxGraphConstantsWithGm)` (not only for wrap), or rebind frozen params/constants on the deserialized graph in `_load_original_gm`.

File: torch/_inductor/output_code.py [hunk 2]
```diff
@@ -968,13 +1103,14 @@ def prepare_for_serialization(self) -> None:
...
         self.compiled_fn_runner = None
+        self._forward_ad_fallback_installed = False
         if self._original_gm is not None:
             from torch.fx._graph_pickler import GraphPickler, Options
 
             self._serialized_original_gm = GraphPickler.dumps(
                 self._original_gm, Options(ops_filter=None)
             )
-            self._original_gm = None
+        self._original_gm = None
```
Comment (bobrenjc93) [inline torch/_inductor/output_code.py]: `prepare_for_serialization` now pickles `_original_gm` for essentially every forward graph, and `GraphPickler.reducer_override` only intercepts `FakeTensor` — real `torch.Tensor`/`Parameter` attributes fall through to default pickling *by value*. That means every FX-graph-cache entry (local and remote) now embeds the real data of every `get_attr` constant, and with freezing the full frozen weights, including CUDA tensors. This defeats the deliberate design in `__init__` (frozen params are kept out of the cache entry and re-looked-up via `frozen_param_names`), and can blow up cache size/write time by orders of magnitude for real models. Please strip/redirect tensor attributes before dumping, or only serialize the fallback graph when something actually needs it.

File: torch/_inductor/output_code.py [hunk 3]
```diff
@@ -715,16 +783,54 @@ def __init__(
...
-            # Store a metadata-stripped copy of the FX graph. Running this
-            # under FakeTensorMode re-derives output shapes and aliasing
-            # from the input fake tensors.
-            import copy
+        # Store a metadata-stripped copy of the FX graph for runtime fallbacks
+        # that need eager FX execution after compilation.  The original gm can
+        # contain large node metadata, so do not keep it alive here.  Backward
+        # graphs do not need the forward AD fallback; only preserve them when
+        # compiled-region wrapping needs a graph for fake tensor propagation.
+        if (
```
Comment (bobrenjc93) [inline torch/_inductor/output_code.py]: This makes the metadata-stripped graph copy unconditional for all forward graphs (previously it was gated behind `wrap_inductor_compiled_regions`). Costs paid by every user on every compile: a `deepcopy` of the whole post-grad graph at compile time (non-trivial for 10k-node graphs), and a permanently retained GraphModule that holds references to all params/buffers/constants — which will keep frozen weights alive and defeat `freezing_discard_parameters`. Please gate this on a config (or build it lazily) so users who never touch forward AD don't pay for it.

File: torch/_inductor/output_code.py [hunk 4]
```diff
@@ -952,10 +1056,41 @@ def wrapped_callable(inputs):
...
 
+        if (
+            self.current_callable is not None
+            and not is_backward
+            and not self._forward_ad_fallback_installed
+        ):
+            fallback_original_callable = self.current_callable
+            self_ref = weakref.ref(self)
+
+            @torch._dynamo.disable(  # type: ignore[misc]
```
Comment (bobrenjc93) [inline torch/_inductor/output_code.py]: This installs an extra Python frame wrapped in `torch._dynamo.disable` around *every* forward Inductor callable. `DisableContext.__call__._fn` does `set_eval_frame(None)` + `_maybe_set_eval_frame(...)`, a `getattr(fn, '__name__')` and `torch.compiler.is_exporting()` per invocation, on top of `_boxed_inputs_have_forward_grad`'s two C calls. That is real per-call latency on the hottest Inductor runtime path (cf. the recent PR that fought for ~3us by removing `detach()` calls), and the benchmark in the PR description only measures cache-hit time, not steady-state call latency. Consider doing the probe inside the existing `CompiledFxGraph.__call__` (no new frame), and/or skipping installation entirely unless a dual level has ever been entered. Also explain why the `disable` is needed at all — if it isn't, drop it.

Comment (jansel) [reply torch/_inductor/output_code.py]: My agent says the unconditional Python wrapper has been removed. Normal execution stays in the existing C++ guard path; the steady-state benchmark measured 11.952 us patched versus 11.909 us baseline, and _fw_grad is queried only while forward AD is active.

File: torch/_inductor/output_code.py (hunk 2 above)
Comment (jansel) [reply torch/_inductor/output_code.py]: My agent says the redesign serializes only boolean guard metadata and no tensor or weight data. All Inductor/AOTAutograd cache-serialization changes were removed, so cache payload size is unchanged by this fix.

File: torch/_inductor/output_code.py (hunk 3 above)
Comment (jansel) [reply torch/_inductor/output_code.py]: My agent says the fallback graph deepcopy and retention have been removed entirely. The final fix is at the Dynamo guard boundary and leaves CompiledFxGraph/cache objects unchanged.

File: torch/_inductor/output_code.py (hunk 1 above)
Comment (jansel) [reply torch/_inductor/output_code.py]: My agent says I replaced the fallback entirely. The fix now invalidates Dynamo graphs when the forward-AD level or tangent presence changes, so dual calls use the existing graph-break/eager-dispatch path. It retains no fallback FX graph or frozen parameters. The new frozen FX-cache regression populates with [2,3,4], hits with [5,6,7], and verifies tangent 18 from the current weights.

**Comment by jansel** [comment]
My agent says I replaced the PR implementation with a root-cause Dynamo guard fix. Forward-AD level changes and primal/dual transitions now invalidate cached graphs, using the existing eager graph-break path to preserve JVPs. This removes fallback graph copying, retention, cache payload, frozen-weight risk, and Python hot-path wrappers. The rebuilt C++ extension, 60 serialization tests, 36 guard-manager tests, focused Dynamo/Inductor regressions, and lintrunner all pass.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says this PR is being closed as part of a cleanup of open agentic PRs to focus review bandwidth on a smaller set; the linked issue remains open.
